File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/VoiceKit/MSCallManager.m
//
//  MSCallManager.m
//  BlackFireIM
//
//  Created by benny wang on 2021/7/20.
//

#import "MSCallManager.h"
#import "MSCallViewController.h"
#import "MSIMSDK-UIKit.h"
#import "AppDelegate.h"

@interface MSCallManager()

@property(nonatomic,copy) NSString *isOnCallingWithUid;

@property(nonatomic,strong) MSCallViewController *callVC;

@property(nonatomic,assign) CallAction action;

@property(nonatomic,assign) MSCallType callType;

@property(nonatomic,strong) NSTimer *timer;

@property(nonatomic,assign) NSInteger timerCount;

@property(nonatomic,copy) NSString *room_id;

@property(nonatomic,assign) BOOL autoAcceptCall;

@end
@implementation MSCallManager

+ (instancetype)shareInstance
{
    static dispatch_once_t onceToken;
    static MSCallManager * g_sharedInstance = nil;
    dispatch_once(&onceToken, ^{
        g_sharedInstance = [[MSCallManager alloc] init];
    });
    return g_sharedInstance;
}

- (void)callToPartner:(NSString *)partner_id
              creator:(NSString *)creator
             callType:(MSCallType)callType
               action:(CallAction)action
              room_id:(nullable NSString *)room_id
{
    if (partner_id.length == 0 || creator.length == 0) return;
    if (room_id) {
        self.room_id = room_id;
    }else {
        NSInteger currentT = [MSIMTools sharedInstance].adjustLocalTimeInterval / 1000 / 1000;
        self.room_id = [NSString stringWithFormat:@"c2c_%@_%zd",[MSIMTools sharedInstance].user_id,currentT];
    }
    switch (action) {
        case CallAction_Call:
        {
            self.isOnCallingWithUid = partner_id;
            self.callType = callType;
            self.callVC = [[MSCallViewController alloc]initWithCallType:self.callType sponsor:creator invitee:partner_id room_id:self.room_id];
            self.callVC.modalPresentationStyle = UIModalPresentationFullScreen;
            AppDelegate *del = (AppDelegate *)[UIApplication sharedApplication].delegate;
            [del.window.rootViewController presentViewController:self.callVC animated:YES completion:nil];
            
            WS(weakSelf)
            self.timerCount = 0;
            [self.timer invalidate];
            self.timer = [NSTimer scheduledTimerWithTimeInterval:1 repeats:YES block:^(NSTimer * _Nonnull timer) {
                
                [weakSelf inviteTimerAction: weakSelf.room_id toReciever:partner_id];
            }];
            [[NSRunLoop mainRunLoop]addTimer:self.timer forMode:NSRunLoopCommonModes];
        }
            break;
        case CallAction_Cancel:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 作为邀请方，取消通话，会补发一条取消通话的普通消息
                [self sendMessageType:CallAction_Cancel option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:self.room_id toReciever:partner_id];
                [self.timer invalidate];
                self.timer = nil;
                self.timerCount = 0;
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_Reject:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id] == NO) {
                /// 作为被邀请方，点击拒绝，结束通话。补一条拒绝的指令消息
                [self sendMessageType:CallAction_Reject option:IMCUSTOM_SIGNAL room_id:self.room_id toReciever:partner_id];
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_End:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 自己是邀请方主动挂断，结束通话同时补发一条结束的普通消息
                [self sendMessageType:CallAction_End option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:self.room_id toReciever:partner_id];
                [self destroyCallVC: room_id];
            }else {
                ///自己是被邀请方主动挂断，结束通话同时补发一条结束的指令消息
                [self sendMessageType:CallAction_End option:IMCUSTOM_SIGNAL room_id:self.room_id toReciever:partner_id];
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_Accept:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id] == NO) {
                /// 作为被邀请方，点击接受，补一条接受的指令消息
                [self.callVC recieveAccept:callType room_id:self.room_id];
                [self sendMessageType:CallAction_Accept option:IMCUSTOM_SIGNAL room_id:self.room_id toReciever:partner_id];
            }
        }
            break;
        default:
            break;
    }
}

- (void)recieveCall:(NSString *)from
            creator:(NSString *)creator
           callType:(MSCallType)callType
             action:(CallAction)action
            room_id:(nullable NSString *)room_id
{
    if (from.length == 0 || [from isEqualToString:[MSIMTools sharedInstance].user_id] || creator.length == 0) return;

    switch (action) {
        case CallAction_Call:
        {
            if (self.isOnCallingWithUid) {
                if (![self.isOnCallingWithUid isEqualToString:from]) {
                    /// 如果正与某人聊天，收到另一邀请指令，会给对方回一条正忙的指令消息
                    [self sendMessageType:CallAction_Linebusy option:IMCUSTOM_SIGNAL room_id:room_id toReciever:from];
                }
                return;
            }
            if ([self.room_id isEqualToString:room_id] == NO) {
                self.room_id = room_id;
                self.isOnCallingWithUid = from;
                self.callType = callType;
                self.callVC = [[MSCallViewController alloc]initWithCallType:self.callType sponsor:from invitee:[MSIMTools sharedInstance].user_id room_id:room_id];
                self.callVC.modalPresentationStyle = UIModalPresentationFullScreen;
                AppDelegate *del = (AppDelegate *)[UIApplication sharedApplication].delegate;
                [del.window.rootViewController presentViewController:self.callVC animated:YES completion:nil];
                if (self.autoAcceptCall) {
                    [self.callVC recieveVoipAccept:callType room_id:room_id];
                }
            }
        }
            break;
        case CallAction_Cancel:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id] == NO) {
                /// 作为被邀请方收到对方取消通话消息，结束通话
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_Timeout:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id] == NO) {
                /// 作为被邀请方，收到对方超时消息，结束通话
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_Reject:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 作为邀请方，收到对方的拒绝指令，结束通话，同时补一条拒绝的普通消息
                [self.timer invalidate];
                self.timer = nil;
                self.timerCount = 0;
                [self sendMessageType:CallAction_Reject option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:room_id toReciever:from];
            }
            [self destroyCallVC: room_id];
        }
            break;
        case CallAction_End:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 作为邀请方，收到对方挂断指令，会结束通话。同时发一条结束的普通消息
                [self sendMessageType:CallAction_End option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:room_id toReciever:from];
                [self destroyCallVC: room_id];
            }else {
                /// 作为被邀请方，收到对方挂断的消息，会结束通话
                [self.callVC recieveHangup:callType room_id:room_id];
                [self destroyCallVC: room_id];
            }
        }
            break;
        case CallAction_Linebusy:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 作为邀请方，收到一条对方正忙的指令，我会结束通话，同时补发一条对方正忙的普通消息
                [self.timer invalidate];
                self.timer = nil;
                self.timerCount = 0;
                [self destroyCallVC: room_id];
                [self sendMessageType:CallAction_Linebusy option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:room_id toReciever:from];
            }
        }
            break;
        case CallAction_Accept:
        {
            if ([creator isEqualToString:[MSIMTools sharedInstance].user_id]) {
                /// 作为邀请方收到对方接受的指令消息
                [self.timer invalidate];
                self.timer = nil;
                self.timerCount = 0;
                [self.callVC recieveAccept:callType room_id:room_id];
            }
        }
            break;
        default:
            break;
    }
}

- (void)inviteTimerAction:(NSString *)room_id toReciever:(NSString *)reciever
{
    [self sendMessageType:CallAction_Call option:IMCUSTOM_SIGNAL room_id:room_id toReciever:reciever];
    /// 作为邀请方，对方60秒内无应答，结束通话。补发一条超时的普通消息
    if (self.timerCount >= 60) {
        [self.timer invalidate];
        self.timer = nil;
        self.timerCount = 0;
        [self destroyCallVC: room_id];
        [self sendMessageType:CallAction_Timeout option:IMCUSTOM_UNREADCOUNT_NO_RECALL room_id:room_id toReciever:reciever];
    }
    self.timerCount++;
}

- (void)autoAcceptVoipCall:(MSCallType)callType room_id:(NSString *)room_id
{
    if (self.callVC) {
        [self.callVC recieveVoipAccept:callType room_id:room_id];
    }else {
        self.autoAcceptCall = YES;
    }
}

- (void)destroyCallVC:(NSString *)room_id
{
    [UIDevice stopPlaySystemSound];
    [self.callVC dismissViewControllerAnimated:YES completion:nil];
    self.callVC = nil;
    self.isOnCallingWithUid = nil;
    self.room_id = nil;
    self.action = CallAction_Unknown;
    self.autoAcceptCall = NO;
    [[NSNotificationCenter defaultCenter]postNotificationName:@"kRecieveNeedToDismissVoipView" object:room_id];
}

- (void)sendMessageType:(CallAction)action option:(MSIMCustomOption)option room_id:(NSString *)room_id toReciever:(NSString *)reciever
{
    NSDictionary *extDic = @{@"room_id": room_id,@"type": @(self.callType == MSCallType_Voice ? MSIMCustomSubTypeVoiceCall : MSIMCustomSubTypeVideoCall),@"event":@(action),@"duration": @(self.callVC.duration)};
    MSIMPushInfo *push;
    if (action == CallAction_Timeout || action == CallAction_Cancel) {
        NSString *attachExt = self.callType == MSCallType_Voice ? @"[Voice call]" : @"[Video call]";
        push = [[MSIMPushInfo alloc]init];
        push.body = [NSString stringWithFormat:@"%@ Call cancelled by caller",attachExt];
        push.sound = @"default";
    }else if (action == CallAction_Call) {
        if (self.timerCount == 0) {
            NSString *attachExt = self.callType == MSCallType_Voice ? @"[Voice call]" : @"[Video call]";
            push = [[MSIMPushInfo alloc]init];
            push.body = [NSString stringWithFormat:@"%@ Start Call",attachExt];
            push.sound = (self.callType == MSCallType_Voice ? @"00.caf" : @"call.caf");
        }
    }else if (action == CallAction_Linebusy || action == CallAction_End) {
        if (option != IMCUSTOM_SIGNAL) {
            NSString *attachExt = self.callType == MSCallType_Voice ? @"[Voice call]" : @"[Video call]";
            push = [[MSIMPushInfo alloc]init];
            push.body = [NSString stringWithFormat:@"%@ Duration: %02zd:%02zd",attachExt,self.callVC.duration/60,self.callVC.duration%60];
            push.sound = @"default";
        }
    }else if (action == CallAction_Reject && option != IMCUSTOM_SIGNAL) {
        NSString *attachExt = self.callType == MSCallType_Voice ? @"[Voice call]" : @"[Video call]";
        push = [[MSIMPushInfo alloc]init];
        push.body = [NSString stringWithFormat:@"%@ Call declined by user",attachExt];
        push.sound = @"default";
    }
    MSIMMessage *message = [[MSIMManager sharedInstance]createVoipMessage:[extDic el_convertJsonString] option:option pushExt:push];
    [[MSIMManager sharedInstance]sendC2CMessage:message toReciever:reciever successed:^(NSInteger msg_id) {
        
            } failed:^(NSInteger code, NSString *desc) {
                MSLog(@"%@",desc);
    }];
}


/// 根据自定义参数解析出消息中展示的内容
+ (NSString *)parseToMessageShow:(NSDictionary *)customParams callType:(MSCallType)callType isSelf:(BOOL)isSelf
{
    CallAction action = [customParams[@"event"] integerValue];
    switch (action) {
        case CallAction_End:
        {
            NSInteger duration = [customParams[@"duration"] integerValue];
            if (callType == MSCallType_Voice) {
                return [NSString stringWithFormat:TUILocalizableString(TUIKitSignalingVoiceCallEnd),duration/60,duration%60];
            }else {
                return [NSString stringWithFormat:TUILocalizableString(TUIKitSignalingVideoCallEnd),duration/60,duration%60];
            }
        }
        case CallAction_Linebusy:
            return isSelf ? TUILocalizableString(TUIKitSignalingCallBusy) : TUILocalizableString(TUIKitSignalingCallBusyOther);
        case CallAction_Timeout:
            return isSelf ? TUILocalizableString(TUIKitSignalingNoResponseOther) : TUILocalizableString(TUIKitSignalingNoResponse);
        case CallAction_Cancel:
            return isSelf ? TUILocalizableString(TUIkitSignalingCancelCall) : TUILocalizableString(TUIkitSignalingCancelCallOther);
        case CallAction_Reject:
            return isSelf ? TUILocalizableString(TUIkitSignalingDeclineOther) : TUILocalizableString(TUIkitSignalingDecline);
        default:
            return TUILocalizableString(TUIkitMessageTipsUnknowMessage);
    }
}

/// 根据自定义参数解析出在会话中展示的内容
+ (NSString *)parseToConversationShow:(NSDictionary *)customParams callType:(MSCallType)callType isSelf:(BOOL)isSelf
{
    NSString *desc = [self parseToMessageShow:customParams callType:callType isSelf:isSelf];
    if (callType == MSCallType_Voice) {
        return [NSString stringWithFormat:@"%@ %@",TUILocalizableString(TUIkitMessageTypeVoiceCall),desc];
    }else {
        return [NSString stringWithFormat:@"%@ %@",TUILocalizableString(TUIkitMessageTypeVideoCall),desc];
    }
}

+ (NSString *)getCreatorFrom:(NSString *)room_id
{
    NSArray *arr = [room_id componentsSeparatedByString:@"_"];
    if (arr.count == 3) {
        return arr[1];
    }
    return @"";
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/CallKit/MSCallManager.swift
//
//  MSCallManager.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/9/28.
//

import UIKit
import MSIMSDK


enum CallAction: Int {
    case error = -1        //系统错误
    case unknown           //未知流程
    case call              //邀请方发起请求
    case cancel            //邀请方取消请求（只有在被邀请方还没处理的时候才能取消）
    case reject            //被邀请方拒绝邀请
    case timeout           //被邀请方超时未响应
    case end               //通话中断
    case linebusy          //被邀请方正忙
    case accept            //被邀请方接受邀请
}

enum MSCallType {
    case voice             //语音通话
    case video             //视频通话
}

enum CallState {
    case dailing           //呼叫
    case onInvitee         //被呼叫
    case calling           //通话中
}

class MSCallManager: NSObject {

    static let shared: MSCallManager = MSCallManager()
    
    func callToPartner(partner_id: String,
                       creator: String,
                       callType: MSCallType,
                       action: CallAction,
                       room_id: String?) {
        if partner_id.count == 0 || creator.count == 0 {return}
        if room_id != nil {
            self.room_id = room_id!
        }else {
            let currentT = MSIMTools.sharedInstance().adjustLocalTimeInterval / 1000 / 1000
            self.room_id = String(format: "c2c_%@_%zd", MSIMTools.sharedInstance().user_id!,currentT)
        }
        switch action {
        case .call:
            self.isOnCallingWithUid = partner_id
            self.callType = callType
            self.callVC = MSCallViewController(callType: self.callType, sponsor: creator, invitee: partner_id, room_id: self.room_id)
            self.callVC?.modalPresentationStyle = .fullScreen
            let appdelegate = UIApplication.shared.delegate as? AppDelegate
            appdelegate?.window?.rootViewController?.present(self.callVC!, animated: true, completion: nil)
            
            self.timerCount = 0
            self.timer?.invalidate()
            self.timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true, block: {[weak self] timer in
                guard let strongSelf = self else{return}
                self?.inviteTimerAction(room_id: strongSelf.room_id,toReciever: partner_id)
            })
            RunLoop.main.add(self.timer!, forMode: .common)
        case .cancel:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 作为邀请方，取消通话，会补发一条取消通话的普通消息
                self.sendMessage(action: .cancel,option: .IMCUSTOM_UNREADCOUNT_NO_RECALL,room_id: self.room_id,toReciever: partner_id)
                self.timer?.invalidate()
                self.timer = nil
                self.timerCount = 0
                self.destroyCallVC()
            }
        case .reject:
            
            if creator != MSIMTools.sharedInstance().user_id {
                /// 作为被邀请方，点击拒绝，结束通话。补一条拒绝的指令消息
                self.sendMessage(action: .reject,option: .IMCUSTOM_SIGNAL,room_id: self.room_id,toReciever: partner_id)
                self.destroyCallVC()
            }
        case .end:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 自己是邀请方主动挂断，结束通话同时补发一条结束的普通消息
                self.sendMessage(action: .end,option: .IMCUSTOM_UNREADCOUNT_NO_RECALL,room_id: self.room_id,toReciever: partner_id)
                self.destroyCallVC()
            }else {
                ///自己是被邀请方主动挂断，结束通话同时补发一条结束的指令消息
                self.sendMessage(action: .end,option: .IMCUSTOM_SIGNAL,room_id: self.room_id,toReciever: partner_id)
                self.destroyCallVC()
            }
        case .accept:
            
            if creator != MSIMTools.sharedInstance().user_id {
                /// 作为被邀请方，点击接受，补一条接受的指令消息
                self.callVC?.recieveAccept(callType: callType, room_id: self.room_id)
                self.sendMessage(action: .accept,option: .IMCUSTOM_SIGNAL,room_id: self.room_id,toReciever: partner_id)
            }
        default:
            break
        }
    }
    
    func recieveCall(from: String,
                     creator: String,
                     callType: MSCallType,
                     action: CallAction,
                     room_id: String?) {
        
        if from.count == 0 || from == MSIMTools.sharedInstance().user_id || creator.count == 0 {return}

        switch action {
        case .call:
            
            if self.isOnCallingWithUid != nil {
                if self.isOnCallingWithUid != from {
                    /// 如果正与某人聊天，收到另一邀请指令，会给对方回一条正忙的指令消息
                    self.sendMessage(action: .linebusy, option: .IMCUSTOM_SIGNAL, room_id: room_id!, toReciever: from)
                }
                return
            }
            if self.room_id != room_id {
                self.isOnCallingWithUid = from
                self.callType = callType
                self.callVC = MSCallViewController(callType: self.callType, sponsor: from, invitee: MSIMTools.sharedInstance().user_id!, room_id: room_id!)
                self.callVC?.modalPresentationStyle = .fullScreen
                let appdelegate = UIApplication.shared.delegate as? AppDelegate
                appdelegate?.window?.rootViewController?.present(self.callVC!, animated: true, completion: nil)
            }
            
        case .cancel:
            
            if creator != MSIMTools.sharedInstance().user_id {
                /// 作为被邀请方收到对方取消通话消息，结束通话
                self.destroyCallVC()
            }
        case .reject:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 作为邀请方，收到对方的拒绝指令，结束通话，同时补一条拒绝的普通消息
                self.timer?.invalidate()
                self.timer = nil
                self.timerCount = 0
                self.sendMessage(action: .reject, option: .IMCUSTOM_UNREADCOUNT_NO_RECALL, room_id: room_id!, toReciever: from)
            }
            self.destroyCallVC()
        case .timeout:
            
            if creator != MSIMTools.sharedInstance().user_id {
                /// 作为被邀请方，收到对方超时消息，结束通话
                self.destroyCallVC()
            }
        case .end:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 作为邀请方，收到对方挂断指令，会结束通话。同时发一条结束的普通消息
                self.sendMessage(action: .end, option: .IMCUSTOM_UNREADCOUNT_NO_RECALL, room_id: room_id!, toReciever: from)
                self.destroyCallVC()
            }else {
                /// 作为被邀请方，收到对方挂断的消息，会结束通话
                self.callVC?.recieveHangup(callType: callType, room_id: room_id!)
                self.destroyCallVC()
            }
        case .linebusy:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 作为邀请方，收到一条对方正忙的指令，我会结束通话，同时补发一条对方正忙的普通消息
                self.timer?.invalidate()
                self.timer = nil
                self.timerCount = 0
                self.destroyCallVC()
                self.sendMessage(action: .linebusy, option: .IMCUSTOM_UNREADCOUNT_NO_RECALL, room_id: room_id!, toReciever: from)
            }
        case .accept:
            
            if creator == MSIMTools.sharedInstance().user_id {
                /// 作为邀请方收到对方接受的指令消息
                self.timer?.invalidate()
                self.timer = nil
                self.timerCount = 0
                self.callVC?.recieveAccept(callType: callType, room_id: room_id!)
            }
        default:
            break
        }
    }
    
    /// 根据自定义参数解析出消息中展示的内容
    class func parseToMessageShow(customParams: [String: Any],callType: MSCallType,isSelf: Bool) -> String? {
        
        guard let actionInt = customParams["event"] as? Int,let action = CallAction(rawValue: actionInt)  else {return nil}
        switch action {
        case .cancel:
            
            return isSelf ? Bundle.bf_localizedString(key: "TUIkitSignalingCancelCall") : Bundle.bf_localizedString(key: "TUIkitSignalingCancelCallOther")
        case .reject:
            
            return isSelf ? Bundle.bf_localizedString(key: "TUIkitSignalingDeclineOther") : Bundle.bf_localizedString(key: "TUIkitSignalingDecline")
        case .timeout:
            
            return isSelf ? Bundle.bf_localizedString(key: "TUIKitSignalingNoResponseOther") : Bundle.bf_localizedString(key: "TUIKitSignalingNoResponse")
        case .end:
            if let duration = customParams["duration"] as? Int {
                if callType == .voice {
                    return String(format: Bundle.bf_localizedString(key: "TUIKitSignalingVoiceCallEnd"), duration / 60, duration % 60)
                }else {
                    return String(format: Bundle.bf_localizedString(key: "TUIKitSignalingVideoCallEnd"), duration / 60, duration % 60)
                }
            }
        case .linebusy:
            
            return isSelf ? Bundle.bf_localizedString(key: "TUIKitSignalingCallBusy") : Bundle.bf_localizedString(key: "TUIKitSignalingCallBusyOther")
        default:
            return Bundle.bf_localizedString(key: "TUIkitMessageTipsUnknowMessage")
        }
        return nil
    }
    
    /// 根据自定义参数解析出在会话中展示的内容
    class func parseToConversationShow(customParams: [String: Any],callType: MSCallType,isSelf: Bool) -> String? {
        
        if let desc = self.parseToMessageShow(customParams: customParams, callType: callType, isSelf: isSelf) {
            if callType == .voice {
                return String(format: "%@ %@", Bundle.bf_localizedString(key: "TUIkitMessageTypeVoiceCall"), desc)
            }else {
                return String(format: "%@ %@", Bundle.bf_localizedString(key: "TUIkitMessageTypeVideoCall"), desc)
            }
        }
        return nil
    }
    
    class func getCreatorFrom(room_id: String) -> String? {
        
        let arr = room_id.components(separatedBy: "_")
        if arr.count == 3 {
            return arr[1]
        }
        return nil
    }
    
    private var isOnCallingWithUid: String?
    
    private var callVC: MSCallViewController?
    
    private var action: CallAction = .error
    
    private var callType: MSCallType = .voice
    
    private var timer: Timer?
    
    private var timerCount: Int = 0
    
    private var room_id: String = ""
    
    private func inviteTimerAction(room_id: String, toReciever: String) {
        
        self.sendMessage(action: .call, option: .IMCUSTOM_SIGNAL, room_id: room_id, toReciever: toReciever)
        /// 作为邀请方，对方60秒内无应答，结束通话。补发一条超时的普通消息
        if self.timerCount >= 60 {
            self.timer?.invalidate()
            self.timer = nil
            self.timerCount = 0
            self.destroyCallVC()
            self.sendMessage(action: .timeout, option: .IMCUSTOM_UNREADCOUNT_NO_RECALL, room_id: room_id, toReciever: toReciever)
        }
        self.timerCount += 1
    }
    
    private func destroyCallVC() {
        if self.callVC != nil {
            UIDevice.stopPlaySystemSound()
            self.callVC?.dismiss(animated: true, completion: nil)
            self.callVC = nil
            self.isOnCallingWithUid = nil
            self.action = .unknown
        }
        NotificationCenter.default.post(name: NSNotification.Name.init("kRecieveNeedToDismissVoipView"), object: self.room_id);
    }
    
    func sendMessage(action: CallAction, option: MSIMCustomOption, room_id: String, toReciever: String) {
        
        let extDic = ["room_id": room_id, "type": (self.callType == .voice ? MSIMCustomSubType.VoiceCall.rawValue : MSIMCustomSubType.VideoCall.rawValue), "event": action.rawValue, "duration": self.callVC!.duration] as [String : Any]
        var push: MSIMPushInfo?
        if action == .timeout || action == .cancel {
            let attachExt = (self.callType == .voice ? "[Voice call]" : "[Video call]")
            push = MSIMPushInfo()
            push?.body = String(format: "%@ Call cancelled by caller", attachExt)
            push?.sound = "default"
        }else if action == .call {
            if self.timerCount == 0 {
                let attachExt = (self.callType == .voice ? "[Voice call]" : "[Video call]")
                push = MSIMPushInfo()
                push?.body = String(format: "%@ Start Call", attachExt)
                push?.sound = (self.callType == .voice ? "00.caf" : "call.caf")
            }
        }else if action == .linebusy || action == .end {
            if option != .IMCUSTOM_SIGNAL {
                let attachExt = (self.callType == .voice ? "[Voice call]" : "[Video call]")
                push = MSIMPushInfo()
                push?.body = String(format: "%@ Duration: %02zd:%02zd", attachExt,self.callVC!.duration / 60,self.callVC!.duration % 60)
                push?.sound = "default"
            }
        }else if action == .reject && option != .IMCUSTOM_SIGNAL {
            let attachExt = (self.callType == .voice ? "[Voice call]" : "[Video call]")
            push = MSIMPushInfo()
            push?.body = String(format: "%@ Call declined by user", attachExt)
            push?.sound = "default"
        }
        let custom = MSIMManager.sharedInstance().createVoipMessage(extDic.bf_convertJsonString(), option: option, pushExt: push)
        MSIMManager.sharedInstance().sendC2CMessage(custom, toReciever: toReciever) { _ in
            
        } failed: { _, desc in
            debugPrint("\(desc ?? "")")
        }
    }
}
